File: netcore/src/Koralium.Data.ArrowFlight/Decoders/BinaryDecoder.cs
/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
using Apache.Arrow;
using Koralium.Data.ArrowFlight.Properties;
using System;
using System.Collections.Generic;
using System.Text;

namespace Koralium.Data.ArrowFlight.Decoders
{
    internal class BinaryDecoder : ColumnDecoder
    {
        private BinaryArray _array;

        public override string GetDataTypeName()
        {
            return "binary";
        }

        public override long GetBytes(in int index, in long dataOffset, in byte[] buffer, in int bufferOffset, in int length)
        {
            if (buffer.Length < (bufferOffset + length))
            {
                throw new InvalidOperationException(Resources.BufferSmallerThanLength);
            }

            var span = _array.GetBytes(index);

            int maxIndex = (int)(span.Length - dataOffset);

            int i = 0;
            
            for (; i < length && i < maxIndex; i++)
            {
                buffer[bufferOffset + i] = span[(int)(i + dataOffset)];
            }

            return i;
        }

        public override Type GetFieldType()
        {
            return typeof(byte[]);
        }

        public override object GetFieldValue(in int index, Type type)
        {
            if(Equals(type, typeof(byte[])))
            {
                return GetValue(index);
            }
            throw new InvalidOperationException($"Cannot cast byte[] to {type.FullName}");
        }

        public override object GetValue(in int index)
        {
            return _array.GetBytes(index).ToArray();
        }

        public override bool IsDbNull(in int index)
        {
            return _array.IsNull(index);
        }

        internal override void NewBatch(IArrowArray arrowArray)
        {
            if (arrowArray is BinaryArray binaryArray)
            {
                _array = binaryArray;
            }
            else
            {
                throw new ArgumentException(Resources.ExpectedBinaryArray, nameof(arrowArray));
            }
        }
    }
}

File: presto/trino-connector-arrowflight/src/main/java/io/trino/plugin/koralium/decoders/BinaryDecoder.java
/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.trino.plugin.koralium.decoders;

import io.airlift.slice.Slice;
import io.airlift.slice.Slices;
import io.trino.spi.block.BlockBuilder;
import io.trino.spi.connector.ConnectorSession;
import io.trino.spi.type.Type;
import org.apache.arrow.vector.VarBinaryVector;
import org.apache.arrow.vector.types.pojo.Field;

import static io.trino.spi.type.VarbinaryType.VARBINARY;

public class BinaryDecoder
        extends PrimitiveDecoder<VarBinaryVector>
{
    @Override
    protected PrimitiveDecoder<VarBinaryVector> createDecoder(Field field, ConnectorSession connectorSession, Type prestoType)
    {
        return new BinaryDecoder();
    }

    @Override
    void writeValue(VarBinaryVector vector, BlockBuilder builder, int index)
    {
        byte[] value = vector.get(index);
        Slice slice = Slices.wrappedBuffer(value);
        VARBINARY.writeSlice(builder, slice);
    }
}
